### PR TITLE
Test: Implemented parallel integration test mechanism

### PR DESCRIPTION
**Pull request by ellistarn**
Fixes # 

**Description**

This test exposes this bug:
```
karpenter-78cc468b57-h689z controller 2022-08-19T23:29:04.324Z	ERROR	controller.provisioning	Could not schedule pod, incompatible with provisioner "markfeather-3-ncysbvhlyj", incompatible requirements, key karpenter.sh/provisioner-name, karpenter.sh/provisioner-name In [arrowrose-4-tzwajhehug] not in karpenter.sh/provisioner-name In [markfeather-3-ncysbvhlyj]; incompatible with provisioner "arrowrose-4-tzwajhehug", key kubernetes.io/hostname, kubernetes.io/hostname DoesNotExist not in kubernetes.io/hostname In [hostname-placeholder-0281]	{"commit": "104576a", "pod": "default/wolverineebony-8-dukkvosoir-586d4c848f-c5qn4"}
```

**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/suites/integration/scheduling_test.go [hunk 1]
```diff
@@ -2,20 +2,142 @@ package integration_test
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type Setup func(provisioner *v1alpha5.Provisioner, nodeTemplate *v1alpha1.AWSNodeTemplate) *appsv1.Deployment
+type Expectation func(map[*v1.Pod]*v1.Node)
+
+var _ = FDescribe("Scheduling", func() {
```
Comment (bwagner5) [inline test/suites/integration/scheduling_test.go:28]:
```suggestion
var _ = Describe("Scheduling", func() {
```

File: test/suites/integration/scheduling_test.go [hunk 2]
```diff
@@ -2,20 +2,142 @@ package integration_test
...
+	It("should schedule pods", func() {
+
+		wg := sync.WaitGroup{}
+		for _, t := range []func() (Setup, Expectation){
+			// WellKnownLabelsTest,
```
Comment (bwagner5) [inline test/suites/integration/scheduling_test.go:33]: ?

File: test/suites/integration/scheduling_test.go [hunk 3]
```diff
@@ -2,20 +2,142 @@ package integration_test
...
+			SelfAffinityTest,
+			SelfAffinityTest,
```
Comment (bwagner5) [inline test/suites/integration/scheduling_test.go:35]: dup?

File: test/suites/integration/scheduling_test.go [hunk 4]
```diff
@@ -2,20 +2,142 @@ package integration_test
...
+			return test.Deployment(test.DeploymentOptions{Replicas: 1, PodOptions: test.PodOptions{
+				NodeSelector:     nodeSelector,
+				NodePreferences:  requirements,
+				NodeRequirements: requirements,
+			}})
+		},
+		func(bindings map[*v1.Pod]*v1.Node) {
+			for pod, node := range bindings {
+				for key, value := range pod.Spec.NodeSelector {
+					Expect(node.Labels).ToNot(HaveKeyWithValue(key, value))
```
Comment (njtran) [inline test/suites/integration/scheduling_test.go:118]: Shouldn't this be the following? Why wouldn't we want the node labels to have the nodeSelectorLabels?
```suggestion
					Expect(node.Labels).To(HaveKeyWithValue(key, value))
```

**Review by njtran — COMMENTED** [review]
One comment, but have high level thoughts: let me know what you think.

Although this was nicely wired together and coded, it took me much more time to create the mental model of how this works than with our other ginkgo test. I'm worried that having an increasingly clever and complex use of ginkgo for parallelism will make creating/reviewing tests more difficult and harder to catch bugs, considering how it all wires up together.  

I'm also slightly concerned about running parallel tests on the same clusters. Bugs with our scheduling code could impact other tests. Differences in expectations between Karpenter and K8s (especially as K8s changes across versions) could also result in weird test flakes. Either of these situations could make diagnosing failures incredibly tough. Even without parallelism, sometimes the cleanup or setup-preconditions fail, which makes me a little more uneasy here. 

WDYT about relying on using multiple clusters for multiple suites and running them in parallel? This ensures CI tests are running in isolated environments, and lowers the risk of our tests impacting each other unexpectedly?

**Comment by njtran** [comment]
Closing for now since tests are running at a good speed.

**Pull request closed without merging**